### PR TITLE
chore(api): remove non-async APIs

### PR DESCRIPTION
**Pull request by lawmicha**
*Issue #, if available:*

*Description of changes:*
This PR removes a bunch of the APIPlugin's non-async APIs and updates the unit tests and integration tests. I understand this PR has a lot of changes, but they're mostly tests that had to be updated, so in terms of reviewing, the commits are in sequential order

- APIPlugin's REST APIs have been fully removed
- GraphQL APIs have been marked as deprecated because it's non-trivial to remove it as it breaks DataStore's usage. DataStore uses API.query for sync query operations and API.mutate for sync to cloud operations. 

**Follow ups**
- API.Subscribe AsyncThrowingSequence will be done as part of another PR in here https://github.com/aws-amplify/amplify-ios/pull/2236
- Remaining API.mutate and API.query will be removed along with DataStore updates together in a follow up.
- Finally, remove API.subscribe non-async API and its usage from DataStore as well.
- Fix warnings in GraphQLConnectionScenario3Tests, ModelBased, and Scenario6 (See comments below)

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Amplify/Categories/API/ClientBehavior/APICategory+GraphQLBehavior.swift [hunk 1]
```diff
@@ -9,6 +9,7 @@ extension APICategory: APICategoryGraphQLBehavior {
     
     // MARK: - Request-based GraphQL operations
 
+    @available(*, deprecated, renamed: "query(request:)")
```
Comment (lawmicha) [inline Amplify/Categories/API/ClientBehavior/APICategory+GraphQLBehavior.swift:12]: These deprecated methods will be removed in another PR, once we get the DataStore implementation updated

File: Amplify/Core/Support/Amplify+Publisher.swift [hunk 2]
```diff
@@ -69,7 +69,7 @@ public extension Amplify {
         ///
         /// - Parameter sequence: The AsyncSequence for which to create the Publisher.
         /// - Returns: The Publisher for the given AsyncSequence.
-        static func create<Sequence: AsyncSequence>(
+        public static func create<Sequence: AsyncSequence>(
```
Comment (lawmicha) [inline Amplify/Core/Support/Amplify+Publisher.swift:72]: had to make this public to create the publisher from the AsyncSequence

File: AmplifyTestCommon/AsyncExpectation.swift [hunk 3]
```diff
@@ -27,6 +27,14 @@ public actor AsyncExpectation {
...
         state == .fulfilled
     }
+    
+    public func setShouldTrigger(_ shouldTrigger: Bool) {
+        self.isInverted = !shouldTrigger
+    }
+    
+    public func setExpectedFulfillmentCount(_ count: Int) {
+        self.expectedFulfillmentCount = count
+    }
```
Comment (lawmicha) [inline AmplifyTestCommon/AsyncExpectation.swift:37]: CC @brennanMKE  I had a bunch of tests that created the expectations in the setUp, then in the test methods control the `isInverted` and `fulfillmentCount`, which were different across test methods. let me know if you have any comments for these changes

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginGraphQLUserPoolTests/Base/AsyncExpectation.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import XCTest
+import Amplify
+
+public actor AsyncExpectation {
```
Comment (lawmicha) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginGraphQLUserPoolTests/Base/AsyncExpectation.swift:12]: Because the HostApp (DataStoreHostApp, APIHostApp, etc) do not have access to AmplifyTestCommon, I am currently copying the AsyncExpectation implementation local to the test target. Had chatted with @ameter before about this but we haven't figured out a better path forward. I could also put it inside DataStoreHostApp app target and then all targets will get it with the `import DataStoreHostApp` too

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift [hunk 5]
```diff
@@ -48,13 +48,13 @@ extension GraphQLConnectionScenario3Tests {
...
-        try await AsyncExpectation.waitForExpectations([connectedInvoked], timeout: TestCommonConstants.networkTimeout)
+        await waitForExpectations([connectedInvoked], timeout: TestCommonConstants.networkTimeout)
         let post = Post3(id: uuid, title: title)
         _ = try await Amplify.API.mutate(request: .create(post))
         let post2 = Post3(id: uuid2, title: title)
         _ = try await Amplify.API.mutate(request: .create(post2))
 
-        try await AsyncExpectation.waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
+        await waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
         await task.cancel()
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:58]: i see a warning `No 'async' operations occur within 'await' expression` at this line in this file. is this something that we can remove?

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift [hunk 6]
```diff
@@ -92,114 +92,95 @@ extension GraphQLConnectionScenario3Tests {
...
-        let disconnectedInvoked = expectation(description: "Connection disconnected")
-        let completedInvoked = expectation(description: "Completed invoked")
-        let progressInvoked = expectation(description: "progress invoked")
-
-        let operation = Amplify.API.subscribe(
-            request: .subscription(of: Post3.self, type: .onDelete),
-            valueListener: { event in
-                switch event {
+    
+    func testOnDeletePostSubscriptionWithModel() async throws {
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:100]: This test seems to run indefinitely on my machine for some reason.

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift [hunk 7]
```diff
@@ -92,114 +91,95 @@ extension GraphQLConnectionScenario3Tests {
         _ = try await Amplify.API.mutate(request: .create(post))
         _ = try await Amplify.API.mutate(request: .update(post))
 
-        try await AsyncExpectation.waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
+        await waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
 
         await task.cancel()
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:96]: Seeing warning message : `No 'async' operations occur within 'await' expression`

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift [hunk 8]
```diff
@@ -92,114 +91,95 @@ extension GraphQLConnectionScenario3Tests {
...
             XCTFail("Failed to create comment with post")
             return
         }
 
-        wait(for: [progressInvoked], timeout: TestCommonConstants.networkTimeout)
-        operation.cancel()
-        wait(for: [disconnectedInvoked, completedInvoked], timeout: TestCommonConstants.networkTimeout)
-        XCTAssertTrue(operation.isFinished)
+        await waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
+        await task.cancel()
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:182]: Seeing warning message : `No 'async' operations occur within 'await' expression`

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario6Tests.swift [hunk 9]
```diff
@@ -59,50 +59,39 @@ class GraphQLConnectionScenario6Tests: XCTestCase {
...
-    func testGetBlogThenFetchPostsThenFetchComments() {
-        guard let blog = createBlog(name: "name"),
-              let post1 = createPost(title: "title", blog: blog),
-              let post2 = createPost(title: "title", blog: blog),
-              let comment1post1 = createComment(post: post1, content: "content"),
-              let comment2post1 = createComment(post: post1, content: "content") else {
+    func testGetBlogThenFetchPostsThenFetchComments() async throws {
+        guard let blog = try await createBlog(name: "name"),
+              let post1 = try await createPost(title: "title", blog: blog),
+              let post2 = try await createPost(title: "title", blog: blog),
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario6Tests.swift:65]: Seeing warning : `Immutable value 'post2' was never used; consider replacing with '_' or removing it`

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests.swift [hunk 10]
```diff
@@ -343,7 +343,7 @@ class GraphQLModelBasedTests: XCTestCase {
         _ = try await Amplify.API.mutate(request: .create(post))
         _ = try await Amplify.API.mutate(request: .update(post))
         
-        try await AsyncExpectation.waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
+        await waitForExpectations([progressInvoked], timeout: TestCommonConstants.networkTimeout)
```
Comment (thisisabhash) [inline AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLModelBasedTests.swift:346]: Seeing some minor warnings in this file 
`No 'async' operations occur within 'await' expression`
`mutate(request:listener:)' is deprecated: renamed to 'mutate(request:)`

**Review by thisisabhash — DISMISSED** [review]

**Review by thisisabhash — APPROVED** [review]

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift (hunk 7 above)
Comment (lawmicha) [reply AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:97]: latest PR to add API.subscribe removes this https://github.com/aws-amplify/amplify-ios/pull/2236

File: AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift (hunk 8 above)
Comment (lawmicha) [reply AmplifyPlugins/API/Tests/APIHostApp/AWSAPIPluginFunctionalTests/GraphQLConnectionScenario3Tests+Subscribe.swift:183]: latest PR to add API.subscribe removes this https://github.com/aws-amplify/amplify-ios/pull/2236
